### DISCUSSION TITLE
How to configure an i2c pwm setup (UI)

### DISCUSSION THREAD

**Discussion by kaedwen** [arendst/Tasmota#17382] [General]
Hi,

I have created a i2c driver for the PCA9632 i2c pwm controller. Therefore I have used the already present PCA9685 driver as a template. Commands like `driver91 enable` or `driver91 pwm,0,255` are functional and the communication is possible.

Now I wonder how to configure the tasmota ui to display 4 PWM sliders. In other RGBW Templates there is a software PWM applied directly to the GPIO's, but how to do it in my case where only GPIO0 (I2C_SCL) and GPIO2 (I2C_SDA) is configured?

And what I have to do that my device will show up in the Homeassistant Tasmota Integration?

Best regards

**Comment by barbudor** [comment]
Hi
I think you do it the wrong way as you first do a PR and then start a discussion on how to do it 😄 

Nevertheless, while I understand that you based your code on the existing PCA9685, may be it's a good opportunity to rethink it
It may be possible to implement "external PWM" to behave like "native PWM" wiht all advantages
For example the PCF8574 IO expander can declare it's output as erlay, making them supporting directly POWER commands

I don't know if this is already feasible for PWM but that could be nice
For that I would suggest to look into light drivers such as `xlgt_02_my92x1.ino`, `xlgt_03_sm16716.ino`, `xlgt_04_sm2135.ino`and others that are drivers for Led drivers / PWM generators used in some light bulbs

**Reply by barbudor** [reply]
Now, a pure PWM driver may have different usage than a Light driver which means that both approach may be valuable

**Reply by kaedwen** [reply]
Thank you for your feedback! Yes you are right there is not that much to discuss about if there is no PR merged ;-)

But yes have found the light drivers in the meanwhile and it looks to be the better way to implement it, so everything is in place to easily get visible in the Homeassistant integration.

Will fight a little bit there

**Comment by kaedwen** [comment]
Started to create a light driver https://github.com/kaedwen/Tasmota/tree/module in `tasmota/tasmota_xlgt_light/xlgt_11_pca9632.ino` but have no idea how to get this linked in the ui? Do I have do define a Module somewhere? And then link the GPIO's to `GPIO_I2C_SCL` and `GPIO_I2C_SDA`?

![image](https://user-images.githubusercontent.com/3512352/210107366-86b03e36-df5c-4c46-b1b0-1857ee832c8b.png)

The log line `LGT: PCA9632 ModuleSelected ... ` is showing up but for some reason the if is not entered. `GPIO_I2C_SCL` and `GPIO_I2C_SDA` are selected in the UI for `GPIO0` and `GPIO2`

The template shows `{"NAME":"Generic","GPIO":[640,0,608,0,0,0,1,1,1,1,1,1,1,1],"FLAG":0,"BASE":18}`

**Reply by kaedwen** [reply]
ok there was a Module other than `General(0)` selected, this was the problem that the `PinUsed` were not truty for some reason

**Reply by kaedwen** [reply]
Have reused the already written raw driver for the PCA_9632 and implemented the light driver around. Working on my side :-D will create a PR for that stuff, have to clean the code a little bit and provide a channel color mapping with some defines.
